specs/phase0/fork-choice.md

```diff
@@ -272,25 +334,87 @@ def get_head(store: Store) -> Root:
+```
 
-    See https://ethresear.ch/t/prevention-of-bouncing-attack-on-ffg/6114 for more detailed analysis and discussion.
+#### `update_unrealized_checkpoints`
+
+```python
+def update_unrealized_checkpoints(store: Store, unrealized_justified_checkpoint: Checkpoint,
+                                  unrealized_finalized_checkpoint: Checkpoint) -> None:
     """
-    if compute_slots_since_epoch_start(get_current_slot(store)) < SAFE_SLOTS_TO_UPDATE_JUSTIFIED:
-        return True
+    Update unrealized checkpoints in store if necessary
+    """
+    # Update unrealized justified checkpoint
+    if unrealized_justified_checkpoint.epoch > store.unrealized_justified_checkpoint.epoch:
+        store.unrealized_justified_checkpoint = unrealized_justified_checkpoint
 
-    justified_slot = compute_start_slot_at_epoch(store.justified_checkpoint.epoch)
-    if not get_ancestor(store, new_justified_checkpoint.root, justified_slot) == store.justified_checkpoint.root:
-        return False
+    # Update unrealized finalized checkpoint
+    if unrealized_finalized_checkpoint.epoch > store.unrealized_finalized_checkpoint.epoch:
+        store.unrealized_finalized_checkpoint = unrealized_finalized_checkpoint
+```
+
+
+#### Pull-up tip helpers
 
-    return True
+##### `compute_pulled_up_tip`
+
+```python
+def compute_pulled_up_tip(store: Store, block_root: Root) -> None:
+    state = store.block_states[block_root].copy()
+    # Pull up the post-state of the block to the next epoch boundary
+    process_justification_and_finalization(state)
+
+    # Store the unrealized justification.
+    store.unrealized_justifications[block_root] = state.current_justified_checkpoint
+
+    # Update unrealized checkpoints in store if necessary
+    update_unrealized_checkpoints(store, state.current_justified_checkpoint, state.finalized_checkpoint)
+
+    # If the block is from a prior epoch, apply the realized values.
+    block_epoch = compute_epoch_at_slot(store.blocks[block_root].slot)
+    current_epoch = compute_epoch_at_slot(get_current_slot(store))
+    if block_epoch < current_epoch:
+        update_checkpoints(store, state.current_justified_checkpoint, state.finalized_checkpoint)
+```
+
+
+#### `on_tick` helpers
+
+##### `on_tick_per_slot`
+
+```python
+def on_tick_per_slot(store: Store, time: uint64) -> None:
+    previous_slot = get_current_slot(store)
+
+    # update store time
+    store.time = time
+
+    current_slot = get_current_slot(store)
+
+    # Reset store.proposer_boost_root if this is a new slot
+    if current_slot > previous_slot:
+        store.proposer_boost_root = Root()
+
+    # Not a new epoch, return
+    if not (current_slot > previous_slot and compute_slots_since_epoch_start(current_slot) == 0):
+        return
+
+    # Pull-up justification and finalization from previous epoch
+    update_checkpoints(store, store.unrealized_justified_checkpoint, store.unrealized_finalized_checkpoint)
```

Comment:
Seems cleaner as an affirmative if and statement (instead of checking the negative and returning)

```suggestion
    # If a new epoch, pull-up justification and finalization from previous epoch
    if  current_slot > previous_slot and compute_slots_since_epoch_start(current_slot) == 0:
        update_checkpoints(store, store.unrealized_justified_checkpoint, store.unrealized_finalized_checkpoint)
```